consul/prepared_query_endpoint.go

```diff
@@ -0,0 +1,581 @@
+		n--
+		i--
+	}
+	return nodes[:n]
+}
+
+// queryServer is a wrapper that makes it easier to test the failover logic.
+type queryServer interface {
+	GetLogger() *log.Logger
+	GetOtherDatacentersByDistance() ([]string, error)
+	ForwardDC(method, dc string, args interface{}, reply interface{}) error
+}
+
+// queryServerWrapper applies the queryServer interface to a Server.
+type queryServerWrapper struct {
+	srv *Server
+}
+
+// GetLogger returns the server's logger.
+func (q *queryServerWrapper) GetLogger() *log.Logger {
+	return q.srv.logger
+}
+
+// GetOtherDatacentersByDistance calls into the server's fn and filters out the
+// server's own DC.
+func (q *queryServerWrapper) GetOtherDatacentersByDistance() ([]string, error) {
+	dcs, err := q.srv.getDatacentersByDistance()
+	if err != nil {
+		return nil, err
+	}
+
+	var result []string
+	for _, dc := range dcs {
+		if dc != q.srv.config.Datacenter {
+			result = append(result, dc)
+		}
+	}
+	return result, nil
+}
+
+// ForwardDC calls into the server's RPC forwarder.
+func (q *queryServerWrapper) ForwardDC(method, dc string, args interface{}, reply interface{}) error {
+	return q.srv.forwardDC(method, dc, args, reply)
+}
+
+// queryFailover runs an algorithm to determine which DCs to try and then calls
+// them to try to locate alternative services.
+func queryFailover(q queryServer, query *structs.PreparedQuery,
+	limit int, options structs.QueryOptions,
+	reply *structs.PreparedQueryExecuteResponse) error {
+
+	// Pull the list of other DCs. This is sorted by RTT in case the user
+	// has selected that.
+	nearest, err := q.GetOtherDatacentersByDistance()
+	if err != nil {
+		return err
+	}
+
+	// This will help us filter unknown DCs supplied by the user.
+	known := make(map[string]struct{})
+	for _, dc := range nearest {
+		known[dc] = struct{}{}
+	}
+
+	// Build a candidate list of DCs to try, starting with the nearest N
+	// from RTTs.
+	var dcs []string
+	index := make(map[string]struct{})
+	if query.Service.Failover.NearestN > 0 {
+		for i, dc := range nearest {
+			if !(i < query.Service.Failover.NearestN) {
+				break
+			}
+
+			dcs = append(dcs, dc)
+			index[dc] = struct{}{}
+		}
+	}
+
+	// Then add any DCs explicitly listed that weren't selected above.
+	for _, dc := range query.Service.Failover.Datacenters {
+		// This will prevent a log of other log spammage if we do not
+		// attempt to talk to datacenters we don't know about.
+		if _, ok := known[dc]; !ok {
+			q.GetLogger().Printf("[DEBUG] consul.prepared_query: Skipping unknown datacenter '%s' in prepared query", dc)
+			continue
+		}
+
+		// This will make sure we don't re-try something that fails
+		// from the NearestN list.
+		if _, ok := index[dc]; !ok {
+			dcs = append(dcs, dc)
+		}
+	}
+
+	// Now try the selected DCs in priority order.
+	failovers := 0
+	for _, dc := range dcs {
+		// This keeps track of how many iterations we actually run.
+		failovers++
```

Comment:
Will `Failovers` be `1` even if the first DC works in this case? The field name was just a little confusing when following the code. If this should be `0` for cases where we never fail over I think you could just use the iterating index instead of incrementing.